### PR TITLE
POC: Try to optimize `columnized_output_exprs` 

### PR DESCRIPTION
**Pull request by goldmedal**
## Which issue does this PR close?

Just a POC for https://github.com/apache/datafusion/issues/13015#issuecomment-2424113317

### CODE REVIEW

File: datafusion/common/src/dfschema.rs [hunk 1]
```diff
@@ -478,6 +479,14 @@ impl DFSchema {
             .collect()
     }
 
+    pub fn column_reference(&self) -> Vec<ColumnReference> {
+        self.iter()
+            .map(|(qualifier, field)| {
+                ColumnReference::new(qualifier.cloned(), field.name().as_str())
+            })
+            .collect()
```
Comment (Dandandan) [inline datafusion/common/src/dfschema.rs:487]: This `Vec` can be avoided by returning iterator?

**Comment by goldmedal** [comment]
The latest commit doesn't make the performance better. Just clean up the codebase.

**Comment by alamb** [comment]
I may be misreading this, but it seems like  columnize_output_expr is called in a loop for each expression being projected:

https://github.com/apache/datafusion/blob/ac827abe1b66b1dfa02ce65ae857477f68667843/datafusion/expr/src/logical_plan/builder.rs#L1505-L1511

However, then for each expr it builds up the same hash table:
https://github.com/apache/datafusion/blob/ac827abe1b66b1dfa02ce65ae857477f68667843/datafusion/expr/src/utils.rs#L879-L883


Maybe we could refactor the code so the hash table is built once rather than once per expression? Something like

```rust
let columizer = Columnizer(plan);
for e in exprs {
  columnizer.columnize(e)
}
```

🤔

**Comment by goldmedal** [comment]
> coupled with your changes to avoid hash map might make a significant difference in planning

Thanks, @alamb. I'll check it in a few days. If it makes sense, I will do the benchmark again 👍

**Comment by goldmedal** [comment]
> However, then for each expr it builds up the same hash table:
> 
> https://github.com/apache/datafusion/blob/ac827abe1b66b1dfa02ce65ae857477f68667843/datafusion/expr/src/utils.rs#L879-L883

It doesn't work for this case because the hash tables are different. The case is 
```
SELECT max(a1), max(a2), ... max(a200) from t
```

I think sharing the hash map may work in some cases like
```
SELECT max(a1), max(a1) + 1, max(a1) +2, ... from t
```
We only need to hash `max(a1)` once, and then look up the column reference many times. In `physical_select_aggregates_from_200` case, we still need to hash every expression.

**Comment by alamb** [comment]
Thanks for double checking @goldmedal

**Pull request closed without merging**